sklearn/inspection/_metric_threshold_curve.py

```diff
@@ -0,0 +1,177 @@
+
+from ..metrics._base import _check_pos_label_consistency
+
+
+@validate_params(
+    {
+        "y_true": ["array-like"],
+        "y_score": ["array-like"],
+        "score_func": [callable],
+        "threshold_grid": [
+            Interval(Integral, 3, None, closed="left"),
+            "array-like",
+            None,
+        ],
+        "pos_label": [Real, str, "boolean", None],
+        "sample_weight": ["array-like", None],
+    }
+)
+def metric_threshold_curve(
+    y_true,
+    y_score,
+    score_func,
+    *,
+    threshold_grid=101,
+    pos_label=None,
+    sample_weight=None,
+):
+    """Compute the threshold-dependent metric of interest per threshold.
+
+    Note: this implementation is restricted to the binary classification task.
+
+    Read more in the :ref:`User Guide <metric_threshold_curve>`.
+
+    .. versionadded:: 1.3
+
+    Parameters
+    ----------
+    y_true : array-like of shape (n_samples,), default=None
+        True targets of binary classification.
+
+    y_score : array-like of shape (n_samples,), default=None
+        Estimated probabilities or output of a decision function.
+
+    score_func : callable
+        Threshold dependent score function (or loss function) with signature
+        `score_func(y, y_pred, sample_weight, **kwargs)`.
+
+    threshold_grid : array-like, int or None, default=101
+        Values of threhsold for each score calculation. If int then
+        `threshold_grid` percentiles of `y_score` are selected. If `None` then
+        all possible thresholds are selected. If int is lower then
+        `len(set(y_score))` then all possible thresholds are selected.
+
+    pos_label : int, float, bool or str, default=None
+        The label of the positive class.
+
+    sample_weight : array-like of shape (n_samples,), default=None
+        Sample weights.
+
+    Returns
+    -------
+    metric_values : ndarray of shape (n_thresholds,)
+        Score value for each threshold. At index i being the value of the
+        theshold-dependent metric for predictions score >= thresholds[i].
+
+    thresholds : ndarray of shape (n_thresholds,)
+        Ascending score values used as thresholds.
+
+    See Also
+    --------
+    precision_recall_curve : Compute precision-recall pairs for different
+        probability thresholds.
+    det_curve : Compute error rates for different probability thresholds.
+    roc_curve : Compute Receiver operating characteristic (ROC) curve.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.metrics import accuracy_score
+    >>> from sklearn.inspection import metric_threshold_curve
+    >>> y_true = np.array([0, 0, 1, 1])
+    >>> y_scores = np.array([0.1, 0.4, 0.35, 0.8])
+    >>> accuracy_values, thresholds = metric_threshold_curve(
+    ...     y_true, y_scores, accuracy_score)
+    >>> thresholds
+    array([0.1 , 0.35, 0.4 , 0.8 ])
+    >>> accuracy_values
+    array([0.75, 0.5 , 0.75, 0.5 ])
```

Comment:
Since this is the file that is triggering the internal error in pytest, I am wondering if it comes from this example.

Could you relaunch the CI by deleting the example?

Reply:
Uhm not really sure why the example was raising the internal error from pytest.

Reply:
I didn't understand either. Super weird cause the literal copy paste of it into a notebook runs the code:

![image](https://user-images.githubusercontent.com/55899543/220167731-3a208829-4f26-4b0a-9e67-3c136e24a8d4.png)

Anyway, it's great that it's working now and we can add an example later. 

Thank you very much @glemaitre!
By the way, this PR is ready for a review now. :)

Reply:
The traceback looks similar to https://github.com/pytest-dev/pytest/issues/6508 but not exactly.

Could you reproduce the failure locally by running the pytest command on your machine? If so, it would be worth trying with the dev version of pytest, and adding some print statements or pdb to inspect what is causing pytest to fail and report the bug to the pytest developers.

Reply:
From https://github.com/scikit-learn/scikit-learn/pull/26091, I am under the impression that we can solve the issue of the example by bumping the `pytest` version (xref: https://github.com/scikit-learn/scikit-learn/pull/26184)

Reply:
I have updated this PR with the most recent changes from the main branch (after #26184 merge), achieving a successful CI status after some quick changes. However, it seems that the error has persisted after I reintroduced the example. :( I'll leave it as is for now so you can see.

I'll go back to it later and report the bug to pytest as ogrisel suggested.

Reply:
Before reporting, let's try to bump to a newer pytest version. In #26184, I only bump to the latest of the 5.X but pytest is already having 7.x.

I will fork branch this PR to make a couple of testing and check if we can resolve it just by updating.

Reply:
I open https://github.com/scikit-learn/scikit-learn/pull/26371 to make some debugging

Reply:
CI is green now! Thank you @glemaitre!